### DISCUSSION TITLE
Issue with `turtlebot4_navigation` scripts on Create 3

### DISCUSSION THREAD

**Discussion by leord22** [iRobotEducation/create3_docs#101] [Questions]
- As far as I know Create 3 documentation only have LIDAR SLAM demo with map generation, but does not have navigation with localization using generated map 
  https://github.com/iRobotEducation/create3_examples/tree/galactic/create3_lidar
- so I installed TurtleBot4 Lite packages on Raspberry PI 4 Ubuntu 20.04 server and remote Desktop portion on virtual machine running Ubuntu Mate 20.04 and ROS2 Galactic based on
  https://turtlebot.github.io/turtlebot4-user-manual/software/turtlebot4_packages.html
  Middleware - Fast-DDS rmw-fastrtps_cpp. Docking, Undocking, manual navigation with teleop_twist_keyboard, RPLIDAR map generating and map saving works perfectly. 
- However, navigation with saved map did not work. I can load saved map with but when I tried to launch navigation on the robot
  `ros2 launch turtlebot4_navigation nav_bringup.launch.py slam:=off localization:=true map:=MyMap.yaml`
  and on the remote desktop:
  `ros2 launch turtlebot4_viz view_robot.launch.py`
  It loads map, but when I tried to execute 2D Pose Estimate, Publish Point and then Nav2 Goal, I only see green arrows (2D pose Estimate, Nav2 Goal) and published coordinates, but nothing else happened on the Rviz2 map, and robot don't move
  
- What are your thoughts about this issue? What else could be misconfigured or missing?
Thank you

**Comment by alsora** [comment]
Hi @leord22, it looks like your problem may be related to the Turtlebot 4 scripts, rather than to the Create 3.

I can try to help you, but I will definitely need more context.
If I was you, I would do the following:
 1. Make sure that desktop, Raspberry Pi and robot can communicate over ROS.
 1. Carefully check the logs from the `turtlebot4_navigation nav_bringup.launch.py` script and make sure that there are no strange warning or logs.
 1. Check what the navigation scripts are doing with command line introspection: for example, when you send a Nav2 goal, do you see the published velocity commands with `ros2 topic echo /cmd_vel` ?

**Reply by leord22** [reply]
Thank you very much Alberto for your reply!
I appreciate your willingness to help and I am sure many Create 3 users will appreciate it too!

1. I am using WI-FI for Create 3 Robot, Raspberry PI and virtual Remote Desktop. They are all on the same subnet and ROS2 communication works. I can see all ROS2 topics on each Raspberry PI and Remote Desktop. I don't use  usb0 networking.

On RASPI robot in each terminal I launched the following
This I sourced for ROS2 and Create3 environment
`source /opt/ros/galactic/setup.bash`

This one I sourced for Turtlebot 4 environment right after
`source ~/turtlebot4_ws/install/local_setup.sh`

RMW implementation
`export RMW_IMPLEMENTATION=rmw_fastrtps_cpp`

2. Started Navigation and my saved map `1_async_home.yaml`
`ros2 launch turtlebot4_navigation nav_bringup.launch.py slam:=off localization:=true map:=1_async_home.yaml`

Here are errors (same in .ros/log)
[amcl-3] [INFO] [1657510913.031172296] [amcl]: Activating
[amcl-3] [INFO] [1657510913.031303201] [amcl]: Creating bond (amcl) to lifecycle manager.
[lifecycle_manager-4] [INFO] [1657510913.181731225] [lifecycle_manager_localization]: Server amcl connected with bond.
[lifecycle_manager-4] [INFO] [1657510913.182312440] [lifecycle_manager_localization]: Managed nodes are active
[lifecycle_manager-4] [INFO] [1657510913.182630843] [lifecycle_manager_localization]: Creating bond timer...
[lifecycle_manager-4] [INFO] [1657510913.182823915] [lifecycle_manager_localization]: Starting managed nodes bringup...
[lifecycle_manager-4] [INFO] [1657510913.182909951] [lifecycle_manager_localization]: Configuring map_server
[map_server-2] [WARN] [1657510913.196430888] [rcl_lifecycle]: No transition matching 1 found for current state active
[map_server-2] [ERROR] [1657510913.196555479] []: Unable to start transition 1 from current state active: Transition is not registered., at /tmp/binarydeb/ros-galactic-rcl-lifecycle-3.1.3/src/rcl_lifecycle.c:355
[lifecycle_manager-4] [ERROR] [1657510913.201236566] [lifecycle_manager_localization]: Failed to change state for node: map_server
[lifecycle_manager-4] [ERROR] [1657510913.201401157] [lifecycle_manager_localization]: Failed to bring up all requested nodes. Aborting bringup.
[controller_server-5] [INFO] [1657510913.497070957] [local_costmap.local_costmap]:
[controller_server-5] local_costmap lifecycle node launched.
[controller_server-5] Waiting on external lifecycle transitions to activate

[planner_server-6] [INFO] [1657510817.551906012] [global_costmap.global_costmap]: Timed out waiting for transform from base_link to map to become available, tf error: Invalid frame ID "map" passed to canTransform argument target_frame - frame does not exist

3. There was nothing published in ros2 topic echo /cmd_vel however it was published the data when I did undocking /docking and teleop_twist_keyboard, so basically as you mentioned and based on the logs, Turtlebot4 navigation does not work.

Would you like me to provide additional information?  Much appreciated. Thanks again!

**Reply by alsora** [reply]
The error
```
[map_server-2] [ERROR] [1657510913.196555479] []: Unable to start transition 1 from current state active: Transition is not registered., at /tmp/binarydeb/ros-galactic-rcl-lifecycle-3.1.3/src/rcl_lifecycle.c:355
```

tells you that you are trying to "configure" (transition 1) a node which is already active.
The lifecycle manager expects to find all nodes in the "unconfigured" state, then call "configure" and then call "activate" if any of these operations return a failure, then the node stops.

You should:
 - check the code that starts the `map_server` node and see make sure that it does not configure it or check that the `lifecycle_manager_localization` is the only component taking care of configuring nodes
 - alternatively, tell the `lifecycle_manager_localization` to not setup the `map_server` since, for some reason, it's already activated.

**Reply by leord22** [reply]
Thank you very much for the troubleshooting! 
I looked through `nav_bringup.launch.py` however I can not find any references to `lifecycle_manager_localization` or `map_server`
I am attaching nav_bringup.launch.py (renamed to nav_bringup.launch.txt) just in case if you would like to look at it
[nav_bringup.launch.txt](https://github.com/iRobotEducation/create3_docs/files/9089732/nav_bringup.launch.txt)

**Comment by jerrynsr** [comment]
Been there,  did that.  Here you can learn from my code:
https://drive.google.com/drive/folders/1kEQjdcMS6RdoPsGjO7-s7uBUfTDM8p5f?usp=sharing

  I created a workspace with irobot create3 and clearpath turtlebot4 code merged.    Once you get that use my scripts to map your house... after the house is mapped use my code to bring up navigator.  Once navigator is up run my initpose script. 
You can then use my rviz2 bringup script to fire up rviz on  a remote computer to verify.   Use rviz to publish the location on the map you want the robot to navigate to.  Use my examples and substitute with your x,y,z info.  

I'm not getting too deep in detail because it's summertime.  I'm outside.. working from home and tonight I do outside stuff.  Also I fly my r/c and fps guided fixed wing drones now.  Also optimizing my robot lawn mower.   The download is huge but it shouldn't take long to have the robot doing what you want.   

When fall arrives I will be coding up more.   My goal is to create a robot rover.  I already ordered my Oak D lite.   I wanted to order it before higher ed buys up all the inventory and there's a supply chain issue.  I already have my create 3 rolling to rooms where my blink cameras detect activity.  I'm using webhooks for that and ROS2 NAV2 with LIDAR to navigate.  Next the OakD camera will be programmed for facial recognition.   If it detect my wife,  my youngest son, or myslef.. it will yip happily.  If it detects my middle or older son, it will growl.
That's the first task anyway.    There's more coming.   

Look at the code.. it's an easy merge.   I posted it shortly before you posted, after that I've been on vacation, I'm back and concentrating on optimizing my robot mower.

**Reply by leord22** [reply]
Thank you very much Jerry for your reply and information!
 I will definitely take a look at your project!
 I think, I've figured it out Clearpath Turtlebot4 SLAM and Navigation part. I am testing it now with different scenarios.
I beleive that my issue was in proper sourcing and correct launching sequence of SLAM and NAV2 Turtlebot4 scripts.

So far, here is SLAM and NAV2 with saved map execution sequence on RASPI that seems to be working:
```
# sourcing ROS2 and Turtlebot4
source /opt/ros/galactic/setup.bash
source ~/turtlebot4_ws/install/setup.sh
# I am using Create 3 recommended Fast-DDS Middleware (installed prior)
export RMW_IMPLEMENTATION=rmw_fastrtps_cpp
# Initiate RPLIDAR
ros2 launch turtlebot4_bringup rplidar.launch.py
# NAV2 - Launching navigation (my_map.yaml is the map name)
ros2 launch turtlebot4_navigation nav_bringup.launch.py slam:=off localization:=true map:=my_map.yaml

```
And on remote desktop:
```

source /opt/ros/galactic/setup.bash
source ~/turtlebot4_ws/install/setup.sh
export RMW_IMPLEMENTATION=rmw_fastrtps_cpp
# starting rviz with the map
ros2 launch turtlebot4_viz view_robot.launch.py
```

I already received Oak-D-Lite camera and it will be my next step to hook it up and run it. My concern is about RASPI handling full load with RPLIDAR and OAK-D camera, it's already very hot without the camera.

When you get a chance, please keep us informed on your project flow and short execution summary / sequence if available. I am also interested in more details about webhooks for ROS2 NAV2 with LIDAR navigation and how it's done. Are you using tornado webserver based app similar to ROSboard?

Thank you very much!
